File: Haskell/src/Screens/Books.hs
module Screens.Books where

import Controllers.Book
import Controllers.Profile
import Data.Char
import DataTypes.Application
import DataTypes.Profile
import Screens.Folder
import Screens.SearchBook
import Utils.Api
import Utils.BookSuggestion
import Utils.Screen

addBookDisplay :: IO String
addBookDisplay = do
  line <-
    putOnScreenCls
      "\n=-=-=-=-=-=-=-=-=-=\nAdd book\n=-=-=-=-=-=-=-=-=-=\n\
      \Enter the name of the book or 'v' to go back:"

  if line == "v"
    then return ""
    else searchBookDisplay line 1

editBookDisplay :: IO String
editBookDisplay = do
  clearScreen
  putStrLn "\n=-=-=-=-=-=-=-=-=-=\nEdit book\n=-=-=-=-=-=-=-=-=-=\n"
  books <- indexBooks
  printBooks books 1

  line <-
    putOnScreen
      "\nChoose an option or digit 'v' to go back:"

  if line == "v"
    then return ""
    else do
      let optionNumber = read line :: Int
      if optionNumber `elem` [1 .. length books]
        then do
          let bookTitle = title $ books !! (optionNumber -1)
          (Just book) <- readBook bookTitle
          enterEditDetails book
          return ""
        else do
          putOnScreen "Invalid option! (Press ENTER to continue)"
          editBookDisplay

enterEditDetails :: Book -> IO String
enterEditDetails book = do
  newRate <- putOnScreen "Enter the new rate: "
  let intRate = read newRate :: Int
  if intRate > 10 || intRate < 0
    then do
      putOnScreen "Invalid rate. The rate must be from 0 to 10. (Press ENTER to continue)"
      enterEditDetails book
    else do
      newDescription <- putOnScreen "Enter the new description: "

      let rateInt = read newRate :: Int
      let newBook = Book (title book) (subject book) (author_name book) rateInt newDescription (folder book) (dateNow book)
      updateBook (title book) newBook

      putOnScreen "Your book has been successfully edited! (Press ENTER to continue)"
      return ""

seeBooksDisplay :: IO String
seeBooksDisplay = do
  clearScreen
  putStrLn "\n=-=-=-=-=-=-=-=-=-=\nList Books\n=-=-=-=-=-=-=-=-=-=\n"
  books <- chooseFolderDisplay
  filteredBooks <- filterBooks books
  if filteredBooks /= []
    then do
      clearScreen
      putStrLn "\n=-=-=-=-=-=-=-=-=-=\nList Books\n=-=-=-=-=-=-=-=-=-=\n"
      printBooks filteredBooks 1
      putOnScreen "\n\n(Press ENTER to continue)"
      return ""
    else do
      return ""

delBookDisplay :: IO String
delBookDisplay = do
  clearScreen
  putStrLn
    "\n=-=-=-=-=-=-=-=-=-=\nDelete book\n=-=-=-=-=-=-=-=-=-=\n"
  books <- indexBooks
  printBooks books 1

  line <-
    putOnScreen
      "\nChoose an option or digit 'v' to go back:"

  if line == "v"
    then return ""
    else do
      let optionNumber = read line :: Int
      if optionNumber `elem` [1 .. length books]
        then do
          let bookTitle = title $ books !! (optionNumber -1)

          successDelete <- deleteBook bookTitle

          putOnScreen "Your book has been successfully deleted! (Press ENTER to continue)"
          updateGoal
          return ""
        else do
          putOnScreen "Invalid option! (Press ENTER to continue)"
          delBookDisplay

editBookGoalDisplay :: IO String
editBookGoalDisplay = do
  line <-
    putOnScreenCls
      "\n=-=-=-=-=-=-=-=-=-=\nEdit reading goal\n=-=-=-=-=-=-=-=-=-=\n\
      \Enter new goal or 'v' to go back:"
  if line == "v"
    then return ""
    else do
      let target = read line :: Int
      books <- indexBooks
      let goal = length books
      let newProfile = Profile goal target
      updateProfile newProfile
      putOnScreen "Your goal has been successfully changed! (Press ENTER to continue)"
      return ""

suggestionDisplay :: IO String
suggestionDisplay = do
  clearScreen
  putStrLn "\n=-=-=-=-=-=-=-=-=-=\nLoading...\n=-=-=-=-=-=-=-=-=-=\n"

  books <- indexBooks

  if null books
    then
      putOnScreenCls
        "\n=-=-=-=-=-=-=-=-=-=\nReading Suggestion\n=-=-=-=-=-=-=-=-=-=\n\
        \You don't have any registered books! \n(Press ENTER to continue)"
    else do
      let mostReadSubject = getMostReadSubject books
      let bestRatedSubject = getBestRatedSubject books

      bestRatedSubjectBooks <- searchSubject bestRatedSubject
      let bothSubjectsBooks = onlyBooksFromSubject bestRatedSubjectBooks mostReadSubject

      if null bothSubjectsBooks
        then
          putOnScreenCls
            "\n=-=-=-=-=-=-=-=-=-=\nReading Suggestion\n=-=-=-=-=-=-=-=-=-=\n\
            \Unfortunately, we don't have enough information about your reading habits! \n(Press ENTER to continue)"
        else do
          chosenBook <- getRandomUnreadBook bothSubjectsBooks

          putOnScreenCls $
            "\n=-=-=-=-=-=-=-=-=-=\nReading Suggestion\n=-=-=-=-=-=-=-=-=-=\n\
            \Based on your readings and ratings:\n\n"
              ++ show chosenBook
              ++ "\n\n(Press ENTER to continue)"
          return ""

filterBooks :: [Book] -> IO [Book]
filterBooks books = do
  option <-
    putOnScreenCls
      "\n=-=-=-=-=-=-=-=-=-=\nFilter\n=-=-=-=-=-=-=-=-=-=\n\
      \Enter 'a' to filter by title, subject or author name\n\
      \or 'r' to filter by minimum rate\n\
      \or 'c' to do not filter\n\
      \Your choice:"
  filterBooksOptions books option

filterBooksOptions :: [Book] -> String -> IO [Book]
filterBooksOptions books option
  | option == "a" = do
    filter <- putOnScreen "Enter the name:"
    return $ search (map toLower filter) books
  | option == "r" = do
    rate <- putOnScreen "Enter the rate:"
    return $ search rate books
  | otherwise = return books

isString :: String -> String -> [String] -> Bool
isString filter actual books
  | actual == filter = True
  | not (null books) = isString filter (head books) (tail books)
  | otherwise = False

search :: String -> [Book] -> [Book]
search filter books = [x | x <- books, isTitle x || isAuthorName x || isSubject x]
  where
    isTitle x = isString filter (head (words (lowerTitle x))) (tail (words (lowerTitle x)))
    isAuthorName x = isString filter (head (lowerAuthorName x)) (tail (lowerAuthorName x))
    isSubject x = isString filter (head (lowerSubject x)) (tail (lowerSubject x))
    lowerTitle book = map toLower (title book)
    lowerAuthorName book = map toLower <$> author_name book
    lowerSubject book = map toLower <$> subject book

File: Prolog/src/screens/books.pl
:- module(screens_books, [screen/1]).
:- encoding(utf8).

:- use_module("./src/screens/main.pl", []).
:- use_module("./src/screens/searchBook.pl", []).

:- use_module("./src/utils/files.pl").
:- use_module("./src/utils/books.pl").
:- use_module("./src/utils/bookSuggestion.pl").
:- use_module("./src/utils/screens.pl").
:- use_module("./src/utils/folders.pl").

:- use_module("./src/controllers/profile.pl").
:- use_module("./src/controllers/books.pl").

screen('add_book'):-
    utils_screens:cls,
    writeln('\n=-=-=-=-=-=-=-=-=-=\nAdd book\n=-=-=-=-=-=-=-=-=-=\n\c
        Enter the name of the book or "v" to go back:'),
    read_line_to_string(user_input, Name),
    (Name == "v" -> screens_main:screen('start'); 
        screens_searchBook:screen('search_book', Name, 1)).

screen('edit_book'):-
    utils_screens:cls,
    writeln('\n=-=-=-=-=-=-=-=-=-=\nEdit book\n=-=-=-=-=-=-=-=-=-=\n'),
    controllers_books:indexBooks(Books),
    utils_books:printBooks(Books, 1),
    writeln('\nChoose an option or digit "v" to go back:'),
    %utils_books:displayBooks,
    %writeln('\nChoose an option or digit "v" to go back:'),
    read_line_to_string(user_input, Choice),
    editOption(Choice, Books).

screen('delete_books'):-
    utils_screens:cls,
    write('\n=-=-=-=-=-=-=-=-=-=\nDelete book\n=-=-=-=-=-=-=-=-=-=\n'),
    controllers_books:indexBooks(Books),
    utils_books:printBooks(Books, 1),
    writeln('\nChoose an option or digit "v" to go back:'),
    read_line_to_string(user_input, Choice),
    deleteOption(Choice, Books).

screen('edit_goal'):-
    utils_screens:cls,
    writeln('\n=-=-=-=-=-=-=-=-=-=\nEdit reading goal\n=-=-=-=-=-=-=-=-=-=\n'),
    writeln('\nEnter new goal or \'v\' to go back:'),
    read_line_to_string(user_input, Choice),
    editGoalOption(Choice).

screen('list_books'):-
    utils_screens:cls,
    write('\n=-=-=-=-=-=-=-=-=-=\nList book\n=-=-=-=-=-=-=-=-=-=\n'),
    controllers_folders:indexFolders(Folders),
    utils_folders:printFolders(Folders, 1),
    writeln('\n Choice the folder or press \'a\' to list all books.'),
    read_line_to_string(user_input, Choice),
    (Choice == "a" -> screen('all_books'); screen('filtered_books', Choice)),
    utils_screens:waitInput.

screen('all_books'):-
    utils_screens:cls,
    write('\n=-=-=-=-=-=-=-=-=-=\nList book\n=-=-=-=-=-=-=-=-=-=\n'),
    utils_books:displayBooks.

screen('book_suggestion'):-
    utils_screens:cls,
    write('\n=-=-=-=-=-=-=-=-=-=\nReading Suggestion\n=-=-=-=-=-=-=-=-=-=\n'),
    writeln('Based on your readings and ratings:\n'),
    utils_bookSuggestion:suggestionBooks,
    utils_screens:waitInput.

screen('filtered_books', Choice):-
    utils_screens:cls,
    utils_folders:getFolderBooks(Choice, Folder),
    write('\n=-=-=-=-=-=-=-=-=-=\nList book\n=-=-=-=-=-=-=-=-=-=\n'),
    utils_books:displayFilteredBooks(Folder).

editOption("v", _):-
    main:screen('start').
editOption(Choice, Books):-
    number_string(Num, Choice),
    %Num >= 1,
    %Num =< 5, % Trocar pelo length
    writeln('Enter the new rate: '),
    read_line_to_string(user_input, NewRate),
    number_string(RateInt, NewRate),
    utils_books:rateValidation(RateInt, Rate),
    writeln('Enter the new description: '),
    read_line_to_string(user_input, NewDescription),
    nth1(Num, Books, ResponseBook),
    NewBook = _{
        'author_name': ResponseBook.author_name,
        'dateNow': ResponseBook.dateNow,
        'description': NewDescription,
        'folder': ResponseBook.folder,
        'rate': Rate,
        'subject': ResponseBook.subject,
        'title': ResponseBook.title
    },
    controllers_books:updateBook(ResponseBook.title, NewBook),
    writeln('\nYour book has been successfully edited!'),
    utils_screens:waitInput.
editOption(_, _):-
    writeln('Invalid option! Try again.'),
    read_line_to_string(user_input, NewChoice),
    editOption(NewChoice).

deleteOption("v", _):-
    main:screen('start').
deleteOption(Choice, Books):-
    number_string(Num, Choice),
    %Num >= 1,
    %Num =< 5, % Trocar pelo length
    utils_books:deleteBook(Books, Num),
    writeln('Your book has been successfully deleted!'),
    utils_screens:waitInput.

deleteOption(_, _):-
    writeln('Invalid option! Try again.'),
    read_line_to_string(user_input, NewChoice),
    deleteOption(NewChoice, _).

editGoalOption("v"):-
    main:screen('start').
editGoalOption(Choice):-
    number_string(NumChoice, Choice),
    integer(NumChoice),
    controllers_profile:updateProfile(Choice),
    writeln('Your goal has been successfully changed!').
editGoalOption(_):-
    writeln('Invalid option! Try again.'),
    read_line_to_string(user_input, NewChoice),
    editGoalOption(NewChoice).